Draft: Ported Telemetry from Opentelemetry-dotnet

stefannikolei opened the pull request:
Added Opentelemetry support #790

I just ported the server Otel code to CoreWCF. I integrated it directly and did not make an extra IDispatchMessageInspector or an IClientMessageInspector. Hope this is ok, but I can also change that

I had no time by now to test if it is really working. Will do that in the coming days.

mconnew commented on src/CoreWCF.Primitives/src/CoreWCF/Telemetry/TelemetryContextMessageProperty.cs:
```diff
@@ -0,0 +1,18 @@
+
+internal sealed class TelemetryContextMessageProperty
+{
+    public const string Name = "OpenTelemetry.Instrumentation.Wcf.Implementation.TelemetryContextMessageProperty";
```
The only important thing with the name is to avoid potential collisions with some other usage. So for example, a name such as "Telemetry" would be bad as multiple libraries could end up using that. The convention I try to stick to is using the fully qualified name of the type as any other code has no legitimate reason to do that. So in this case, it would be `CoreWCF.Telemetry.TelemetryContextMessageProperty`. It's not used for anything that's sent over the wire, it's all local to the running process so in the end, it really doesn't matter, as long as you avoid potential collision/clashes with app code and what that would be doing.

mconnew commented on src/CoreWCF.Primitives/src/CoreWCF/Telemetry/WcfInstrumentationEventSource.cs:
```diff
@@ -0,0 +1,89 @@
+namespace CoreWCF.Telemetry;
+
+
+[EventSource(Name = "OpenTelemetry-Instrumentation-Wcf")]
```
The EventSource name is used to derive the ETW Guid when running on Windows. If you use the same name between the OTel WCF library and here, anything listening for these events would get events from OTel with WCF and CoreWCF. If the event message format diverges, that might cause a problem. I think some of the dotnet cross platform command line event collection tooling would also be exposed to this name. I think this should probably be unique.

stefannikolei replied:
renamed it

mconnew commented on src/CoreWCF.Primitives/src/CoreWCF/Telemetry/HttpRequestMessagePropertyWrapper.cs:
```diff
@@ -0,0 +1,114 @@
+
+/// <summary>
+/// This is a reflection-based wrapper around the HttpRequestMessageProperty class. It is done this way so we don't need to
+/// have an explicit reference to System.ServiceModel.Http.dll. If the consuming application has a reference to
```
Fix up references to System.ServiceModel in comments

mconnew commented on src/CoreWCF.Primitives/src/CoreWCF/Telemetry/HttpRequestMessagePropertyWrapper.cs:
```diff
@@ -0,0 +1,114 @@
+        try
+        {
+            type = Type.GetType(
+                "System.ServiceModel.Channels.HttpRequestMessageProperty, System.ServiceModel, Version=0.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089",
```
These type names need to be updates to use CoreWCF namespaces and the correct assembly name

mconnew commented on src/CoreWCF.Primitives/src/CoreWCF/Telemetry/PropertyFetcher.cs:
```diff
@@ -0,0 +1,123 @@
+    /// <param name="obj">Object to be fetched.</param>
+    /// <param name="value">Fetched value.</param>
+    /// <returns><see langword="true"/> if the property was fetched.</returns>
+    public bool TryFetch(object obj, out T value)
```
I understand what's trying to be done here. There's some problems with the code, eg there's no check that the inner fetcher is for the same type as obj. I think we might be able to avoid the need for this though. We add the HttpContext as a message property (we use the fully qualified class name as the key), and that is a resolvable type in Primitives. You can read the headers from there without dealing with any reflection.

stefannikolei commented:
This is the information you get when instrumented
![image](https://github.com/user-attachments/assets/3b477083-600b-4a56-a007-694a59009a47)
![image](https://github.com/user-attachments/assets/5f83da26-87a7-4a1c-b40c-9f0d8e7dfe03)

stefannikolei commented:
One thing I do not know is the reference to Opentelemetry Nuget package. Is that ok?

After a first review I would tackle some tests for this. But I could probably take some help how to test it. Just some direction or if there are some tests I can look at and figure it out.

mconnew commented:
I need to look at what Aspire/aspnetcore does to support OT as they don't reference the OpenTelemetry NuGet package as far as I'm aware. It might be the case that we need to output to a dotnet runtime built in api which would be consumed by some other code which then calls OpenTelemetry. I'm not opposed to having a new CoreWCF.OpenTelemetry package which makes things easy for the consumer, but I'm pretty sure directly depending on the OpenTelemetry packages is probably the wrong thing to do.  
@samsp-msft, what are your thoughts about this? Do you know what api's aspnetcore calls for OTel integration?

mconnew commented on src/CoreWCF.Primitives/src/CoreWCF/Telemetry/AssemblyVersionExtensions.cs:
```diff
@@ -0,0 +1,34 @@
+
+#if NET || NETSTANDARD2_1_OR_GREATER
+        var indexOfPlusSign = informationalVersion!.IndexOf('+', StringComparison.Ordinal);
+#else
```
If you look at the overload for `String.IndexOf(char value, StringComparison comparisonType)` [here](https://github.com/dotnet/runtime/blob/dae890906431049d32e24d498a1d707a441a64a8/src/libraries/System.Private.CoreLib/src/System/String.Searching.cs#L61), you can see that passing `StringComparison.Ordinal` just calls the overload `String.IndexOf(char value)`. This conditional compilation block isn't needed.

mconnew commented on src/CoreWCF.Primitives/src/CoreWCF/Telemetry/AssemblyVersionExtensions.cs:
```diff
@@ -0,0 +1,34 @@
+        // The following parts are optional: pre-release label, pre-release version, git height, Git SHA of current commit
+        // For package version, value of AssemblyInformationalVersionAttribute without commit hash is returned.
+
+        var informationalVersion = assembly.GetCustomAttribute<AssemblyInformationalVersionAttribute>()?.InformationalVersion;
```
Whenever possible, I prefer to use GetCustomAttributesData to get the attribute. It doesn't require instantiating the attribute itself.

stefannikolei replied:
Do you think it is worth changing it? Should it be faster, or should we benchmark this for the best solution?

mconnew commented on src/CoreWCF.Primitives/src/CoreWCF/Telemetry/TracerProviderBuilderExtensions.cs:
```diff
@@ -0,0 +1,45 @@
+    public static TracerProviderBuilder AddWcfInstrumentation(this TracerProviderBuilder builder,
+        Action<WcfInstrumentationOptions>? configure)
+    {
+        if (WcfInstrumentationActivitySource.Options != null)
```
This isn't necessary if we use the configuration options pattern. As long as `WcfInstrumentationOptions` has a default constructor (which it does), you can inject `IOptions<WcfInstrumentationOptions>` options somewhere useful and fetch its `.Value` property. This will trigger DI to create an instance of `WcfInstrumentationOptions` using the default constructor, then run any configuration delegates that have been registered against it, then return the value from the `.Value` property. Basically you can rewrite this method like this:
```c#
public static TracerProviderBuilder AddWcfInstrumentation(this TracerProviderBuilder builder,
        Action<WcfInstrumentationOptions>? configure)
{
    builder.ConfigureServices(services =>
    {
        if (configure != null)
        {
            services.Configure(configure);
        }
    });

    return builder.AddSource(WcfInstrumentationActivitySource.ActivitySourceName);
}
```
There's still the matter of setting the options, but I'll follow up in a more general comment for that. I'm looking at the aspnetcore instrumentation and it looks like there's a separation of concerns between aspnetcore and the OTel aspnetcore instrumentation library that we should probably copy. Basically "in-box" (meaning Primitives), it sends to a DiagnosticSource the raw primitives (e.g. HttpContext) at certain events. Then the "out of box" (which would be e.g. a new CoreWCF.OpenTelemetry package) handles working out what you want to send etc and calls methods on ActivitySource to actually emit the data. I need to understand things a bit better and I need to get the next CoreWCF release out before I do that. These are thoughts in progress, so don't commit too hard on a direction until I've distilled my ideas a bit more.

samsp-msft replied:
>I'm looking at the aspnetcore instrumentation and it looks like there's a separation of concerns between aspnetcore and the OTel aspnetcore instrumentation library that we should probably copy. Basically "in-box" (meaning Primitives), it sends to a DiagnosticSource the raw primitives (e.g. HttpContext) at certain events. Then the "out of box" (which would be e.g. a new CoreWCF.OpenTelemetry package) handles working out what you want to send etc and calls methods on ActivitySource to actually emit the data.

We are trying to move away from that model in .NET! Now that semantic conventions are stabilizing, we want the instrumentation to be built in - its one less thing to drag around, and it means that implicit instrumentation and tooling like VisualStudio will be able to show telemetry, even if the app author didn't explicitly enable it.

For example, https://github.com/dotnet/runtime/pull/103922/files#diff-1486d9b1b5c7729c3633a4f5c1a07226d6d0fe76f5e9732601c21d4fb4cf5c9c Adds an activity directly to HttpClient.

stefannikolei commented:
> I need to look at what Aspire/aspnetcore does to support OT as they don't reference the OpenTelemetry NuGet package as far as I'm aware. It might be the case that we need to output to a dotnet runtime built in api which would be consumed by some other code which then calls OpenTelemetry. I'm not opposed to having a new CoreWCF.OpenTelemetry package which makes things easy for the consumer, but I'm pretty sure directly depending on the OpenTelemetry packages is probably the wrong thing to do. @samsp-msft, what are your thoughts about this? Do you know what api's aspnetcore calls for OTel integration?

Yeah I knew that directly depending on Otel would be wrong. Just wanted to get through and get some feedback and a direction how to do it ;) SO I am waiting to proceed on it :) 

When we have this sorted out I still have to look into testing this. Probably the EndpointTest is a good starting point for testing this Otel stuff, or?

g7ed6e commented:
I think we should be able to remove the OTEL dependencies using Activities and exposing the ActivitySource name. User code would then be responsible to add the activity source name to the OTEL TracingBuilder.

stefannikolei commented:
> I think we should be able to remove the OTEL dependency using Activities and exposing the ActivitySource name. User code would then be responsible to add the activity source name to the OTEL TracingBuilder.

So I delete the builder code and make  the ActivitySourceName public?

samsp-msft commented:
> I need to look at what Aspire/aspnetcore does to support OT as they don't reference the OpenTelemetry NuGet package as far as I'm aware. It might be the case that we need to output to a dotnet runtime built in api which would be consumed by some other code which then calls OpenTelemetry. I'm not opposed to having a new CoreWCF.OpenTelemetry package which makes things easy for the consumer, but I'm pretty sure directly depending on the OpenTelemetry packages is probably the wrong thing to do. @samsp-msft, what are your thoughts about this? Do you know what api's aspnetcore calls for OTel integration?

I haven't drilled into the PR yet, but off the top of my head, there should not need to be any direct dependencies on OTel libs - they should pick it up from your named ActivitySource.

samsp-msft commented:
> > I think we should be able to remove the OTEL dependency using Activities and exposing the ActivitySource name. User code would then be responsible to add the activity source name to the OTEL TracingBuilder.
> 
> So I delete the builder code and make the ActivitySourceName public?
> 
> Then it should be in the docs how to enable it

Yes this is the correct pattern.

For example, in an Aspire ServiceDefaults you would do 

``` c#
           .WithTracing(tracing =>
           {
               tracing.AddSource(builder.Environment.ApplicationName)
                   .AddAspNetCoreInstrumentation()
                   .AddHttpClientInstrumentation()
                   .AddSource("CoreWCF.Something")
           });
```

With the name being published as part of the docs. We are trying to move away from instrumentation libraries to built-in so that things like implicit instrumentation can work.

You can have multiple names if applicable. .NET networking added some extras in .NET 9 for DNS, sockets, TLS and HttpConnection creation. These are useful for troubleshooting, but are probably too verbose for 24x7 use in production, so you add them as extra sources when needed.

samsp-msft commented on src/CoreWCF.Primitives/src/CoreWCF/Telemetry/WcfInstrumentationEventSource.cs:
```diff
@@ -0,0 +1,60 @@
+namespace CoreWCF.Telemetry;
+
+
+[EventSource(Name = "OpenTelemetry-Instrumentation-CoreWCF")]
```
This is a bit of a philosophical question:- should these be ETW events which may be monitored on Windows, but will probably never be seen in a linux/container environment? The alternative would be to use ILogger and have them collected that way.

stefannikolei replied:
I can easily remove them. I would not need those. I would just remove it, when it's ok. Those events are just for Filtering and Enrichment.

samsp-msft commented on src/CoreWCF.Primitives/src/CoreWCF/Telemetry/WcfInstrumentationOptions.cs:
```diff
@@ -0,0 +1,40 @@
+    /// If Filter returns true, the request is collected.
+    /// If Filter returns false or throw exception, the request is filtered out.
+    /// </summary>
+    public Func<Message, bool>? IncomingRequestFilter { get; set; }
```
We don't do this for other scenarios - filtering is typically enabled after OTel by something like the OTel collector.

samsp-msft commented on src/CoreWCF.Primitives/src/CoreWCF/Telemetry/WcfInstrumentationOptions.cs:
```diff
@@ -0,0 +1,40 @@
+    /// <para>object: the raw <see cref="Message"/> from which additional information can be extracted to enrich the activity.
+    /// </para>
+    /// </remarks>
+    public Action<Activity, string, object>? Enrich { get; set; }
```
We don't yet have a pattern for a callback API for enrichment as part of the runtime for other Activity events. Instrumentation libraries such as https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Instrumentation.AspNetCore include that as a capability.

Maybe a more ideal pattern would be:
- The Core of CoreWCF doesn't take a dependency on OTel - it just uses activity, with well published names
- There is a helper library that includes the configuration portion, and a nicer Tracing.AddWcfTelemetry() helper
- The helper library includes events/callbacks for further enrichment
  - Typically these not only pass the activity, but also domain specific objects that they may want to pull the data from. In Http, it passes the HttpContext. Is there an equivalent for CoreWCF?

stefannikolei replied:
Can we just drop it for the first round and add this capability when it is really needed?

mconnew replied:
I agree to adding the capability when needed, but as an FYI, the equivalent of HttpContent is OperationContext

samsp-msft commented on src/CoreWCF.Primitives/src/CoreWCF/Telemetry/WcfInstrumentationConstants.cs:
```diff
@@ -0,0 +1,19 @@
+    public const string RpcSystemTag = "rpc.system";
+    public const string RpcServiceTag = "rpc.service";
+    public const string RpcMethodTag = "rpc.method";
+    public const string NetHostNameTag = "net.host.name";
```
Are these from the OTel Semantic conventions? I see `network.peer.address` and `network.peer.port` as the way RPC & Http define these.

https://github.com/open-telemetry/semantic-conventions/blob/main/docs/rpc/rpc-spans.md

If there are additional WCF specific ones needed, then we should start a conversation with @lmolkova.

stefannikolei replied:
I just reused those from the Otel community packages. If you want them to be changed let me know

samsp-msft replied:
Please match names to the semantic conventions - this is the common standard from OTel for naming span properties, metrics etc. I think most are correct, its the `net.*` that needs to change.

stefannikolei replied:
done

stefannikolei commented:
I now have addressed some of your feedback. Especially removing the Otel dependencie.

Besides that I stripped down the implementation:
- Removed ETW Events
- Removed ability to Filter Requests
- Removed ability for enrichment

I would suggest to go for the minimal implementation and then add those features when needed.

stefannikolei commented:
@mconnew @samsp-msft I think i am for my part done now. I added some tests. If you need something more, then let me know.

stefannikolei commented:
> @mconnew @samsp-msft I think i am for my part done now. I added some tests. If you need something more, then let me know.

Ok I have got a problem with this test. The tests run in parallel and my listener now gets the activities of all tests run in parallel. Does anyone have an idea how to avoid that?

The only thing I can imagine is to create a testproject which just has this test.

schaeferrobert commented:
Hey, when is this function coming?

sherland commented:
Which release are you targeting this for?

g7ed6e commented on src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/ImmutableDispatchRuntime.cs:
```diff
@@ -660,6 +682,50 @@ internal async Task<MessageRpc> ProcessMessageAsync(MessageRpc rpc)
+
+                    var actionMetadata = GetActionMetadata(request, action);
+
+                    activity.SetTag(WcfInstrumentationConstants.RpcServiceTag, actionMetadata.ContractName);
```
whould we set `RpcServiceTag` and `RpcMethodTag` with null values ?

stefannikolei replied:
'RpcServiceTag' could be null --> should probably only write it when set,
'RpcMethodTag' should not be null --> string.Empty, or the Action from the request Header. So should normally be set in most cases

g7ed6e reviewed (approved):
LGTM.

mconnew commented:
After a bunch of time spent getting up to speed on Aspire (sticking point was automatically wiring up the endpoint took a bit of work to understand the mechanics for non-HttpClient transports), I finally got an end-to-end solution working with this. I don't have answers yet, but there are a few things I need to spend a bit more time thinking about as they aren't quite right. Feel free to suggest alternatives.

The source name, I don't like using "CoreWCF.Primitives". I think in the future we're going to have multiple tiers of sources for different purposes, with different levels of info being emitted. E.g. it would be nice to dump the entire SOAP body (if that's even possible?), or at least all the SOAP headers (need to be smart about redacting information here). I need to explore Aspire more, as some of it might be more applicable to structured logs, or events. That's out of scope for this PR, but just giving you an idea of where I want to take this. So with that in mind, I want a name which basically means just the bare minimum basic CoreWCF info.

When looking at the list of items in the Span section of the traces, the value for soap.message_version seems way too long. It's showing as `Soap11 (http://schemas.xmlsoap.org/soap/envelope/) AddressingNone (http://schemas.microsoft.com/ws/2005/05/addressing/none)`. The name prefix of soap implies this isn't WCF specific, so other soap stacks might emit this. But the value has some WCF-isms in it. Soap11 and AddressingNone are the names of the corresponding static properties of EnvelopeVersion and AddressingVersion (which make up the MessageVersion). You wouldn't expect to have the same values for non-WCF stacks. We should either have it be wcf.message_version and just use `Soap11,AddressingNone` as the value, or we should just have the soap spec namespaces as part of the value: `http://schemas.xmlsoap.org/soap/envelope/,http://schemas.microsoft.com/ws/2005/05/addressing/none`. Is this copied from the existing WCF extensibility or something you came up with?

Is there any documentation around the meaning of the Name field? It's showing as the name of soap action that was used. That seems redundant with rpc.method. I'm wondering if the name should be something like the name of the service implementation type.  

I still need to go through the PR implementation to see if there's anything that I would like addressing, but I suspect that will be minimal. I'm going to activate the CoPilot code reviewer as that will likely pick up most of the small things I might mention.

mconnew commented:
@stefannikolei, it looks like there's an effort to overhaul the rpc spec (which I'll be getting involved with now I know about it). In [the blog post announcing it](https://opentelemetry.io/blog/2025/stabilizing-rpc-conventions/#towards-reliable-telemetry-conventions), it says the migration plan to new semantics would be to have any differences behind a flag. I don't think it would serve the community to have 3 different ways of doing things, the existing WCF extensibility way, the CoreWCF way, and the eventual standardized way. So for now we'll keep as strictly to the existing WCF extensibility as possible to keep any churn that's needed by consumers to a minimum. I think the easiest way to switch between two semantics would be two different source names. So let's keep this set of spans to using `"CoreWCF.Primitives"` and any future work will use a different source name.

mconnew commented:
@stefannikolei, the current build is breaking because one of the typos you fixed in a name wasn't fixed in all the places it's used. The point of usage is still using the misspelt name. Also can you rebase these changes and this PR can't be merged until you have.

mconnew commented on src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/ImmutableClientRuntime.cs:
```diff
@@ -13,7 +14,7 @@ internal class ImmutableClientRuntime
         private readonly Dictionary<string, ProxyOperationRuntime> _operations;
-
+        private readonly Dictionary<string, ActionMetadata> _actionMappings;
         internal ImmutableClientRuntime(ClientRuntime behavior)
```
nit: should have a blank line between fields and constructor.

mconnew commented on src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/ImmutableClientRuntime.cs:
```diff
@@ -29,14 +30,22 @@ internal ImmutableClientRuntime(ClientRuntime behavior)
+                    operationName: clientOperation.Name);
+            }
+
+
```
nit: there's an extra blank line here, there should only be one

mconnew commented on src/CoreWCF.Primitives/src/CoreWCF/Telemetry/ActionMetadata.cs:
```diff
@@ -0,0 +1,17 @@
+    }
+
+    public string? ContractName { get; set; }
+
```
Strings are nullable as they are references. We haven't done the work to enable nullability in this code base so no need to specify.

mconnew commented on src/CoreWCF.Primitives/src/CoreWCF/Telemetry/AssemblyVersionExtensions.cs:
```diff
@@ -0,0 +1,31 @@
+        Debug.Assert(!string.IsNullOrEmpty(informationalVersion), "AssemblyInformationalVersionAttribute was not found in assembly");
+
+        var indexOfPlusSign = informationalVersion!.IndexOf('+');
+        
```
The ! signifying this isn't null isn't needed.

stefannikolei commented on src/CoreWCF.Primitives/src/CoreWCF/Telemetry/TelemetryContextMessageProperty.cs:
```diff
@@ -0,0 +1,18 @@
+    public TelemetryContextMessageProperty(IDictionary<string, ActionMetadata> actionMappings)
+    {
+        ActionMappings = actionMappings;
+    }
```
I would leave it like that. in the current usage it is always created

mconnew commented:
@stefannikolei, would you be interested in writing a blog post with instructions on how to use this with Aspire? If you don't have the bandwidth, I'll put something together and ask you to review it.

stefannikolei commented:
@mcconnew would you mind pushing a preview version to Nuget? That would greatly simplify setting up a demo.

stefannikolei commented:
@mconnew Would you mind into looking into the build failures. https://github.com/CoreWCF/CoreWCF/runs/43566134641

The last Nuget package on the private feed is alpha.2

It should be alpha.18 by now

mconnew commented:
@sherland, I appreciate you chiming in. The problem was the .NET Foundation switched from service accounts to managed identities for the code signing. It's taken a bit of work to get things setup right, but we've got it working now.  

@stefannikolei, there's a newer version available published to the dev feed now which includes the open telemetry changes. Sorry about the delay getting this fixed. Let me know if you need anything from me for the blog post.
